op-node/rollup/derive/batch.go

```diff
@@ -7,48 +7,44 @@ import (
 	"io"
 	"sync"
 
-	"github.com/ethereum-optimism/optimism/op-node/rollup"
-	"github.com/ethereum-optimism/optimism/op-service/eth"
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
 // Batch format
 // first byte is type followed by bytestring.
 //
-// BatchV1Type := 0
-// batchV1 := BatchV1Type ++ RLP([epoch, timestamp, transaction_list]
-//
 // An empty input is not a valid batch.
 //
 // Note: the type system is based on L1 typed transactions.
-
+//
 // encodeBufferPool holds temporary encoder buffers for batch encoding
 var encodeBufferPool = sync.Pool{
 	New: func() any { return new(bytes.Buffer) },
 }
 
 const (
-	BatchV1Type = iota
+	// SingularBatchType is the first version of Batch format, representing a single L2 block.
+	SingularBatchType = iota
+	// SpanBatchType is the Batch version used after SpanBatch hard fork, representing a span of L2 blocks.
+	SpanBatchType
 )
 
-type BatchV1 struct {
-	ParentHash common.Hash  // parent L2 block hash
-	EpochNum   rollup.Epoch // aka l1 num
-	EpochHash  common.Hash  // block hash
-	Timestamp  uint64
-	// no feeRecipient address input, all fees go to a L2 contract
-	Transactions []hexutil.Bytes
+// Batch contains information to build one or multiple L2 blocks.
+// Batcher converts L2 blocks into Batch and writes encoded bytes to Channel.
+// Derivation pipeline decodes Batch from Channel, and converts to one or multiple payload attributes.
+type Batch interface {
+	GetBatchType() int
+	GetTimestamp() uint64
+	LogContext(log.Logger) log.Logger
 }
 
+// BatchData is a composition type that contains raw data of each batch version.
+// It has encoding & decoding methods to implement typed encoding.
 type BatchData struct {
-	BatchV1
-	// batches may contain additional data with new upgrades
-}
-
-func (b *BatchV1) Epoch() eth.BlockID {
-	return eth.BlockID{Hash: b.EpochHash, Number: uint64(b.EpochNum)}
+	BatchType int
+	SingularBatch
+	RawSpanBatch
```

Comment:
Adding the type explicitly, and embedding each kind, is probably the less intrusive change we can make to the existing approach with typed BatchData. We can maybe refactor it to be more similar to `types.Transaction` in upstream geth, where a struct is used to represent the typed decoding/encoding and any caching, and wraps around a single interface, rather than embedding each type of possible inner batch. We can keep this for now, to keep the PR stack easier to land into develop.

Reply:
Will do this in a separate PR. Strongly agree that following what `types.Transaction` did is better because current composition pattern will become more complicated when we need another sub field of composition. We need to add additional field for new span batch type when needed if we do not use a single interface.

Let me make the abstraction idea concrete. type `BatchData` will be refactored as below:
```go
type BatchData struct {
	inner BatchInnerData
}
```
where `BatchInnerData` is an interface:
```go
type BatchInnerData interface {
	batchType() int
        encode(w io.Writer) error
        decode(r *bytes.Reader) error
}
```

Each `SingularBatch` and `RawSpanBatch` will implement `BatchInnerData` interface. This will encapsulate all the batch type related conditionals to each implementations: for example below method will be refactored from:
```go
func (b *BatchData) encodeTyped(buf *bytes.Buffer) error {
	switch b.BatchType {
	case SingularBatchType:
		buf.WriteByte(SingularBatchType)
		return rlp.Encode(buf, &b.SingularBatch)
	case SpanBatchType:
		buf.WriteByte(byte(b.BatchType))
		return b.RawSpanBatch.encode(buf)
	default:
		return fmt.Errorf("unrecognized batch type: %d", b.BatchType)
	}
}
```
to
```go
func (b *BatchData) encodeTyped(buf *bytes.Buffer) error {
        if err := buf.WriteByte(b.inner.batchType()); err != nil {
            return err
        }
        return b.inner.encode(buf)
}
```

Reply:
I just found out that `buf.WriteByte(SingularBatchType)` is not error handled. Will also fix this.